### PR TITLE
Optionals

### PR DESCRIPTION
**Pull request by kevin1e100**
proposed optional extensions (KT-50484)

### CODE REVIEW

**Review by ilya-g — CHANGES_REQUESTED** [review]

File: libraries/stdlib/jdk8/src/kotlin/jvm/optionals/Optionals.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+package kotlin.jvm.optionals
+
+import java.util.Optional
+
+/**
+ * Returns this [Optional]'s value if [present][Optional.isPresent], or otherwise `null`.
+ */
+public fun <T : Any> Optional<out T>.getOrNull(): T? = orElse(null)
```
[the remark this answers is not in the document]
Comment (kevin1e100) [reply libraries/stdlib/jdk8/src/kotlin/jvm/optionals/Optionals.kt:13]: Seemingly `out` would make this function technically more similar to something like `List.firstOrNull()`, since `List`'s type parameter is declared `out` but `Optional`'s is not, IIUC (https://youtrack.jetbrains.com/issue/KT-49210). So in the `toList()` case for instance, `out` seems to be needed to allow something like `Optional.of("foo").toList<CharSequence>()`.

But I agree it doesn't seem to have practical value here, so I removed `out` except in the `toXxx` functions, where it seems to enable some corner cases (see above), for which I added tests. Let me know if you want me to remove `out` (and the extra tests) for `toXxx` as well, though.

File: libraries/stdlib/jdk8/test/jvm/optionals/OptionalsTest.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        assertNull(Optional.empty<String>().getOrDefault(null))
+        assertEquals(5.0, Optional.empty<Int>().getOrDefault<Number, Int>(5.0))
+    }
+
+    @Test
+    fun getOrElse() {
+        assertEquals("foo", Optional.of("foo").getOrElse { throw AssertionError() })
+        assertEquals("bar", Optional.empty<String>().getOrElse { "bar" })
+
+        assertFailsWith<IllegalStateException>(message = "ran") {
```
[the remark this answers is not in the document]
Comment (kevin1e100) [reply libraries/stdlib/jdk8/test/jvm/optionals/OptionalsTest.kt:34]: Done, thanks!

File: libraries/stdlib/jdk8/src/kotlin/jvm/optionals/Optionals.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Returns this [Optional]'s value if [present][Optional.isPresent], or otherwise [defaultValue].
+ */
+public fun <R, T : R & Any> Optional<out T>.getOrDefault(defaultValue: R): R = if (isPresent) get() else defaultValue
```
[the remark this answers is not in the document]
Comment (kevin1e100) [reply libraries/stdlib/jdk8/src/kotlin/jvm/optionals/Optionals.kt:18]: Done

**Review by ilya-g — APPROVED** [review]
